### PR TITLE
Enable Docker CI

### PR DESCRIPTION
**Pull request by AshwinHIBM**
The purpose of each job:

1. periodic-config-docker: Check the configuration of the kernel for suitability to run Docker.
2. periodic-build-dev-image-docker: Build a minimal Docker development image.
3. postsubmit-unit-test-docker: Run unit tests against https://github.com/moby/moby
4. postsubmit-integration-test-docker: Run integration tests against https://github.com/moby/moby
5. postsubmit-integration-cli-docker: Run the end-to-end test suite from https://github.com/docker/cli

### CODE REVIEW

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 1]
```diff
@@ -0,0 +1,207 @@
+periodics:
```
Comment (Rajalakshmi-Girish) [inline config/jobs/periodic/docker/periodic-ci-docker.yaml:1]: I hope this PR restores the CI jobs that were part of [#447](https://github.com/ppc64le-cloud/test-infra/pull/447).

Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:1]: Yes

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        - base_ref: main
+          org: ppc64le-cloud
+          repo: docker-ce-build
+      reporter_config:
+          slack:
+            channel: 'prow-job-notifications'
+            report_template: 'Job {{.Spec.Job}} of type {{.Spec.Type}} ended with state {{.Status.State}}. <!subteam^S02N6DWBX0F> <{{.Status.URL}}|View logs>'
+      spec:
+        containers:
+        - image: quay.io/powercloud/docker-ce-build@sha256:b89afc315f96f5cfd43de82b3b02f9221983592b893f838d816db5954088c9fc
```
Comment (Rajalakshmi-Girish) [inline config/jobs/periodic/docker/periodic-ci-docker.yaml:18]: We typically version the images in `quay.io` and use them in Prow jobs. Using version tags here would make this line cleaner and more maintainable.

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+          args:
+          - -c
+          - |
+            set -o errexit
+            set -o nounset
+            set -o pipefail
+            set -o xtrace
+            echo "* Start prow-info-docker *"
+            chmod ug+x $PWD/upstream-master-ci/prow-info-docker.sh
+            $PWD/upstream-master-ci/prow-info-docker.sh
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:35]: The others will still be attempted as they're all independent of each other by design. I can see what you're implying; if there's such a problem with the underlying cluster such that we can't even query the kernel for basic information, there's no point trying to run the other jobs. What can the effects of failure ideally be?

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            set -o pipefail
+            set -o xtrace
+            echo "* Start prow-integration-tests *"
+            chmod ug+x $PWD/upstream-master-ci/prow-integration-tests.sh
+            $PWD/upstream-master-ci/prow-integration-tests.sh
+            rc=$?
+            [ $rc != 0 ] && echo "ERROR: prow-integration-tests exited with code:$rc"
+            exit $rc
+          securityContext:
+            privileged: true
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:207]: The [lint job failure](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/pr-logs/pull/ppc64le-cloud_test-infra/517/pull-test-infra-yamllint/1866778956489822208) indicates that I'm missing a new line (I think the hyphen indicates that I don't have a blank line at the end of the file since it doesn't have a line number before it). Should I add a newline?

[the remark this answers is not in the document]
Comment (Rajalakshmi-Girish) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:207]: Yes try adding a new line. I sometimes encounter EOF-related lint errors when editing YAML files using inappropriate editors.
To avoid such issues, it’s best to use a proper YAML-friendly editor that ensures consistent formatting. 

If needed, you can always rely on the pull-test-infra-yamllint job to catch and fix YAML formatting issues. For example, refer to this configuration:
[pull-test-infra-yamllint Job Configuration](https://github.com/ppc64le-cloud/test-infra/blob/master/config/jobs/ppc64le-cloud/test-infra/test-infra-presubmits.yaml#L3:L15).

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rajalakshmi-Girish) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:35]: I’m unsure about the purpose of the upstream-master-ci/prow-info-docker.sh script or what exactly it runs. Could you please determine the intended job functionality and assess the potential impact of its failure?

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 2 above)
Comment (Rajalakshmi-Girish) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:18]: I’m not fully convinced yet. Perhaps we could discuss this further offline?

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            set -o nounset
+            set -o pipefail
+            set -o xtrace
+            echo "* Start prow-build-dev-image *"
+            chmod ug+x $PWD/upstream-master-ci/prow-build-dev-image.sh
+            $PWD/upstream-master-ci/prow-build-dev-image.sh
+            rc=$?
+            [ $rc != 0 ] && echo "ERROR: prow-build-dev-image exited with code:$rc"
+            exit $rc
+          securityContext:
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:85]: These properties are a part of `periodics[*].spec.containers[*]`.

config.yaml:
```
- labels:
    preset-build-docker: "true"
  resources:
    requests:
      cpu: "4000m"
    limits:
      cpu: "4000m"
  securityContext:
    privileged: true
```
doesn't get read. Is there another job where such a nested property has been grouped to compare with?

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+      reporter_config:
+        slack:
+          channel: 'prow-job-notifications'
+          report_template: 'Job {{.Spec.Job}} of type {{.Spec.Type}} ended with state {{.Status.State}}. <!subteam^S02N6DWBX0F> <{{.Status.URL}}|View logs>'
+      spec:
+        containers:
+        - image: quay.io/powercloud/docker-ce-build@sha256:b89afc315f96f5cfd43de82b3b02f9221983592b893f838d816db5954088c9fc
+          resources:
+            requests:
+              cpu: "8000m"
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml]: 4000ms each should work. Does that seem ok?

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+          command:
+          - /bin/bash
+          args:
+          - -c
+          - |
+            set -o errexit
+            set -o nounset
+            set -o pipefail
+            set -o xtrace
+            echo "* Start prow-integration-tests *"
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:159]: Interesting observation!

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            rc=$?
+            [ $rc != 0 ] && echo "ERROR: prow-integration-tests exited with code:$rc"
+            exit $rc
+          securityContext:
+            privileged: true
+    - name: periodic-integration-test-docker
+      cluster: k8s-ppc64le-cluster
+      labels:
+        preset-build-docker: "true"
+      cron: 30 14 * * *
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:171]: No logic. I used 2 p.m. and 2:30 p.m., two slots that you suggested. They're independent of each other, so I can modify this schedule if necessary.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:35]: Sorry, please let me rephrase. I can say that this job's failure guarantees failure of the other 3 jobs. Given that, are you suggesting that we should have a way to "cancel" the remaining jobs' execution if this fails?

**Comment by clnperez** [comment]
@Rajalakshmi-Girish has some good comments/requests so I'll approve if she does. Thanks Raji!

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rajalakshmi-Girish) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:85]: You are right. This isn't possible.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 2 above)
Comment (Rajalakshmi-Girish) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:18]: Discussed offline.
https://ibm-systems-power.slack.com/archives/G0125USRSG1/p1735932311027679?thread_ts=1735284429.102429&cid=G0125USRSG1

**Comment by AshwinHIBM** [comment]
[One approach](https://github.com/ppc64le-cloud/test-infra/pull/517/commits/fb05c557f1659ca34bd7e004ccb09abd51380ed0) for
https://github.com/ppc64le-cloud/test-infra/pull/517#discussion_r1879675626
[doesn't work](https://s3.us-south.cloud-object-storage.appdomain.cloud/ppc64le-prow-logs/pr-logs/pull/ppc64le-cloud_test-infra/517/pull-test-infra-prow-checkconfig/1886725290420342784/build-log.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=d5b0d1e12bb74ec5b0cca25bec3e0746%2F20250204%2Fus-south%2Fs3%2Faws4_request&X-Amz-Date=20250204T105413Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&x-id=GetObject&X-Amz-Signature=c8ed58e59b7a43bc394cb65c33cf221f40d841ccf356763659e8ec352c4eb5fc), dropping.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml]: Thanks, adjusted the periodics to 00:00 and 00:10.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:46]: Converted unit tests, integration tests and integration-cli tests from periodics to postsubmits since I drew a conclusion that if the kernel configuration isn't suited to run Docker, there's no point running the other tests, they will fail.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+      cron: 0 3 * * *
+      decorate: true
+      extra_refs:
+        - base_ref: main
+          org: ppc64le-cloud
+          repo: docker-ce-build
+      reporter_config:
+        slack:
+          channel: 'prow-job-notifications'
+          report_template: 'Job {{.Spec.Job}} of type {{.Spec.Type}} ended with state {{.Status.State}}. <!subteam^S02N6DWBX0F> <{{.Status.URL}}|View logs>'
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:83]: One approach to address this by moving `reporter_config` to config.yaml
[doesn't work](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/pr-logs/pull/ppc64le-cloud_test-infra/517/pull-test-infra-prow-checkconfig/1886725290420342784), dropping.

**Comment by mkumatag** [comment]
@Rajalakshmi-Girish lets review this PR

**Comment by Rajalakshmi-Girish** [comment]
> @Rajalakshmi-Girish lets review this PR

I shall review this in a day.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            # Set the SSH key for git from the secret 'git-ssh-docker-ce-build'
+            echo "* Set up the ssh key for git *"
+            mkdir -p ~/.ssh
+            cp /etc/ssh-key/ssh-privatekey  ~/.ssh/id_rsa
+            chmod 600 ~/.ssh/id_rsa
+            ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts
+
+            echo "* Start prow-info-docker *"
+            chmod ug+x $PWD/upstream-master-ci/prow-info-docker.sh
+            $PWD/upstream-master-ci/prow-info-docker.sh
```
Comment (Rajalakshmi-Girish) [inline config/jobs/periodic/docker/periodic-ci-docker.yaml:46]: @AshwinHIBM You mentioned `periodic-config-docker` is to  `Check the configuration of the kernel for suitability to run Docker.`, where as line here https://github.com/ppc64le-cloud/docker-ce-build/blob/main/upstream-master-ci/prow-info-docker.sh#L14 seems to display a text `Prow Job to run CI tests on the Docker packages`, Seems some discrepancy.

Comment (Rajalakshmi-Girish) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:46]: I see DATE being directed to `/home/prow/go/src/github.com/ppc64le-cloud/docker-ce-build/env/date.list` at https://github.com/ppc64le-cloud/docker-ce-build/blob/main/upstream-master-ci/prow-info-docker.sh#L12
But there doesn't seem to be a file date.list in https://github.com/ppc64le-cloud/docker-ce-build/tree/main/env!
@AshwinHIBM Can we connect to understand better?

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      labels:
+        preset-build-docker: "true"
+      cron: 0 0 * * *
+      decorate: true
+      extra_refs:
+        - base_ref: main
+          org: ppc64le-cloud
+          repo: docker-ce-build
+      reporter_config:
+        slack:
```
Comment (Rajalakshmi-Girish) [inline config/jobs/periodic/docker/periodic-ci-docker.yaml:13]: I came across https://github.com/kubernetes-sigs/prow/blob/main/site/content/en/docs/components/core/crier.md#slack-reporter

Seems like we can move all the slack `reporter_config:` section in jobs to https://github.com/ppc64le-cloud/test-infra/blob/master/config/prow/config.yaml#L5 mentioning the `org/repo:` (ppc64le-clou/docker-ce-build:) like in below example:
image

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]
@AshwinHIBM Let us schedule a call tomorrow. Please help me understand better.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml (hunk 10 above)
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:39]: > @AshwinHIBM You mentioned periodic-config-docker is to Check the configuration of the kernel for suitability to run Docker., where as line here https://github.com/ppc64le-cloud/docker-ce-build/blob/main/upstream-master-ci/prow-info-docker.sh#L14 seems to display a text Prow Job to run CI tests on the Docker packages, Seems some discrepancy.

https://github.com/ppc64le-cloud/docker-ce-build/pull/271

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]
Do all jobs `postsubmit-unit-test-docker`, `postsubmit-integration-test-docker`, `postsubmit-integration-test-docker` require dind image?

Also, here are my comments for your scripts at https://github.com/ppc64le-cloud/docker-ce-build/tree/main/upstream-master-ci - just sharing my thoughts!
- Why the mention of `adding a git commit` at https://github.com/ppc64le-cloud/docker-ce-build/blob/main/setup-pj-trigger.sh#L20? I see only date info being added to folder under https://github.com/ppc64le-cloud/docker-ce-build/tree/prow-job-tracking/job
- Why the intermediate https://github.com/ppc64le-cloud/docker-ce-build/blob/main/setup-pj-trigger.sh? Why not directly call https://github.com/ppc64le-cloud/docker-ce-build/blob/main/trigger-prow-job-from-git.sh from https://github.com/ppc64le-cloud/docker-ce-build/blob/main/upstream-master-ci/prow-info-docker.sh
- How is moby dynamic here? https://github.com/ppc64le-cloud/docker-ce-build/blob/main/upstream-master-ci/build-dev-image.sh#L1 Where is it cloned?
- Why two files `prow-unit-test-docker.sh` `unit-tests.sh`? Can't the common part from files of first kind be clubbed at one place?

**Comment by AshwinHIBM** [comment]
Squashing https://github.com/ppc64le-cloud/test-infra/pull/517/commits/7dece6446baefaacf9919b12b6df83a0989be18d and https://github.com/ppc64le-cloud/test-infra/pull/517/commits/114c05319de4586c416c9fec3330fb7ff21e978e.

File: config/jobs/periodic/docker/periodic-ci-docker.yaml [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+            # Set the SSH key for git from the secret 'git-ssh-docker-ce-build'
+            echo "* Set up the ssh key for git *"
+            mkdir -p ~/.ssh
+            cp /etc/ssh-key/ssh-privatekey  ~/.ssh/id_rsa
+            chmod 600 ~/.ssh/id_rsa
+            ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts
+
+            echo "* Start prow-info-docker *"
+            chmod ug+x $PWD/upstream-master-ci/prow-info-docker.sh
+            $PWD/upstream-master-ci/prow-info-docker.sh
```
[the remark this answers is not in the document]
Comment (AshwinHIBM) [reply config/jobs/periodic/docker/periodic-ci-docker.yaml:39]: cgroups, apparmor, etc. Are you suggesting we remove it from CI and only run it when something changes with the underlying cluster like the OS version? I'm fine with that idea (we probably don't need to run it daily) but is there a way to know something changed with the cluster? Or instead changing the frequency to once a week/month?

**Comment by Rajalakshmi-Girish** [comment]
@AshwinHIBM Do we still need this? Should we revive this change and re review?
